### PR TITLE
More documentation organization + fix `offset1` mess.

### PR DESCRIPTION
**Pull request by Tokazama**
* Doc changes: We may want to add more subgroups to these traits, types, or function sections; but I figured we should make trait methods easier to find as soon as possible before splitting hairs of more detailed organization here.
* Test changes: fixed deprecation of `Val`-> `ColumnNorm` argument in `qr` method.
* Code changes: I also figured I'd fix the mess I've created with `offset1` here and consolidate the `OffsetArrays` code.

### CODE REVIEW

File: test/runtests.jl [hunk 1]
```diff
@@ -53,7 +53,7 @@ rowind,colind=findstructralnz(Sp)
 
 
 @test !fast_scalar_indexing(qr(rand(10, 10)).Q)
-@test !fast_scalar_indexing(qr(rand(10, 10), Val(true)).Q)
+@test !fast_scalar_indexing(qr(rand(10, 10), ColumnNorm()).Q)
```
Comment (chriselrod) [inline test/runtests.jl:56]: Want to put this behind a `if VERSION >= v"1.7"` check?

**Review by chriselrod — APPROVED** [review]

**Comment by chriselrod** [comment]
This PR broke LoopVectorization's tests.
I should really check more thoroughly before approving things.

**Comment by Tokazama** [comment]
Can you link me to that. I'm not seeing it in the actions.

**Comment by chriselrod** [comment]
Locally.

**Comment by chriselrod** [comment]
The failures are going to be related to:
```julia
julia> using OffsetArrays, ArrayInterface

julia> A = OffsetArray(rand(10,10), 100, 200);

julia> ArrayInterface.offsets(A)
(101, 201)

julia> ArrayInterface.offset1(A)
static(1)

julia> ArrayInterface.known_offset1(A)
1
```

**Comment by Tokazama** [comment]
Did we want to keep multidimensional arrays having `offset1` equal 1?

**Comment by chriselrod** [comment]
One failure is probably related to:
```julia
julia> using OffsetArrays, ArrayInterface

julia> A = OffsetArray(rand(10,10), 100, 200);

julia> ArrayInterface.offsets(A)
(101, 201)

julia> ArrayInterface.known_offset1(view(A,105,:))
1

julia> view(A,105,:)[1]
ERROR: BoundsError: attempt to access 10-element view(OffsetArray(::Matrix{Float64}, 101:110, 201:210), 105, :) with eltype Float64 with indices 201:210 at index [1]
Stacktrace:
 [1] throw_boundserror(A::SubArray{Float64, 1, OffsetMatrix{Float64, Matrix{Float64}}, Tuple{Int64, Base.Slice{OffsetArrays.IdOffsetRange{Int64, Base.OneTo{Int64}}}}, true}, I::Tuple{Int64})
   @ Base ./abstractarray.jl:691
 [2] checkbounds
   @ ./abstractarray.jl:656 [inlined]
 [3] getindex(V::SubArray{Float64, 1, OffsetMatrix{Float64, Matrix{Float64}}, Tuple{Int64, Base.Slice{OffsetArrays.IdOffsetRange{Int64, Base.OneTo{Int64}}}}, true}, i::Int64)
   @ Base ./subarray.jl:308
 [4] top-level scope
   @ REPL[10]:1

julia> view(A,105,:)[205]
0.7838614161434267
```

**Comment by chriselrod** [comment]
I'll have to spend more time investigating. I made a couple (now deleted) comments that turned out not to be fully correct.

**Comment by Tokazama** [comment]
Sorry about this. It's after midnight my time, but I'll check back in the morning and see what I can do.

**Comment by Tokazama** [comment]
Let me know if this is an issue and needs my attention still. I'd hate to leave you cleaning up after my mess here.

**Comment by chriselrod** [comment]
It's still an issue. I have upper bounds on ArrayInterface compatibility for VectorizationBase and LoopVectorization at the moment.

**Comment by chriselrod** [comment]
I spent a little time looking at it
ArrayInterface v3.1.23:
```julia
julia> A = OffsetArray(view(rand(100,100), 50:59, 30:39), 100, 1000);

julia> av = view(A, 106, :);

julia> stridedpointer(av)
VectorizationBase.StridedPointer{Float64, 1, -1, -1, (2,), Tuple{Int64}, Tuple{Int64}}(Ptr{Float64} @0x0000000005dc2010, (800,), (1001,))

julia> stridedpointer(OffsetArray(av, OffsetArrays.Origin(0)))
VectorizationBase.StridedPointer{Float64, 1, -1, -1, (2,), Tuple{Int64}, Tuple{Int64}}(Ptr{Float64} @0x0000000005dc2010, (800,), (0,))
```
ArrayInterface master:
```julia
julia> A = OffsetArray(view(rand(100,100), 50:59, 30:39), 100, 1000);

julia> av = view(A, 106, :);

julia> stridedpointer(av)
VectorizationBase.StridedPointer{Float64, 1, -1, -1, (2,), Tuple{Int64}, Tuple{Int64}}(Ptr{Float64} @0x0000000005be0d50, (80,), (1001,))

julia> stridedpointer(OffsetArray(av, OffsetArrays.Origin(0)))
VectorizationBase.StridedPointer{Float64, 1, -1, -1, (2,), Tuple{StaticInt{8}}, Tuple{Int64}}(Ptr{Float64} @0x0000000005be0d50, (static(8),), (0,))
```

**Comment by Tokazama** [comment]
The issue is we aren't specifically calling `strides(A::OffsetArray) = strides(parent(A))` I thought we were already accounting for parent arrays. Should we be doing:

```julia
function strides(x::X) where {X}
    if parent_type(X) <: X
      if defines_strides(x)
          return size_to_strides(size(x), One())
      else
          return Base.strides(x)
      end
    else
      return strides(parent(x))
    end
end
```
or require that each type explicitly define a strides method?

**Comment by chriselrod** [comment]
> The issue is we aren't specifically calling `strides(A::OffsetArray) = strides(parent(A))` I thought we were already accounting for parent arrays. Should we be doing:
> 
> ```julia
> function strides(x::X) where {X}
>     if parent_type(X) <: X
>       if defines_strides(x)
>           return size_to_strides(size(x), One())
>       else
>           return Base.strides(x)
>       end
>     else
>       return strides(parent(x))
>     end
> end
> ```
> 
> or require that each type explicitly define a strides method?

I think this makes sense as the default. I think there are relatively few array wrappers changing strides (like `PermutedDimsArray`) compared to those that don't.

**Pull request merged**